### PR TITLE
fix: support for Result<T, E> with exception throwing on Dart

### LINKED ISSUE #533: `Result<(),MyCustomError>` instead of `anyhow`?
This library is very useful, but having to return a text error instead of a nice struct with custom error information would be much nicer! Is it possible? Is it in the plans?

### PR DESCRIPTION
**Pull request by SiongSng**
## Changes
Continue #582
Fixes #533
Fixes #1070

## Checklist

- [x] An issue to be fixed by this PR is listed above.
- [x] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [x] The code generator is run and the code is formatted (via `just precommit`).
- [x] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [x] CI is passing.

## Remark for PR creator

- Justfile is a task runner for the command line interface that allows you to run shell commands from a file named `justfile` in your project directory. You can use Justfile after [installing it](https://github.com/casey/just). Note that commands written in `justfile` of this repository are expected to be run in `bash`, not `cmd` or `powershell`. Running `just ...` commands in `cmd` or `powershell` will produce errors as the syntax is not compatible. On Windows, you can use `git bash` if you have Git installed.
- If fzyzcjy does not reply for a few days, maybe he just did not see it, so please ping him.

### CODE REVIEW

**Comment by SiongSng** [comment]
The part I'm currently struggling with is ([mod.rs](https://github.com/fzyzcjy/flutter_rust_bridge/blob/55e09b3015d40fbac35651469dd474214173d067/frb_codegen/src/parser/mod.rs#L189)), I can't figure out how to merge between the newly refactored code by(#1117) and the result returning code.

@coder0xff any ideas?

**Comment by lattice0** [comment]
I really need this feature, but I don't have the time to finish it. I hope you do well, thank you very much!

**Comment by fzyzcjy** [comment]
@lattice0 Also thanks for your code for the feature!

**Comment by JustSimplyKyle** [comment]
Now new [issue](https://github.com/fzyzcjy/flutter_rust_bridge/blob/5c2ed40ba108bc8b11a8679d2579721f8659e328/frb_codegen/src/parser/ty.rs#L343)
`let enu = self.parse_enum_core(&p.ident, p.is_exception);`
`p` is the `SupportedPathType`  which got refactored out
`p.ident` should be just the name, but I can't find a way to replace `p.is_exception`

**Comment by JustSimplyKyle** [comment]
I've ran into a weird issue
after this [commit ](https://github.com/fzyzcjy/flutter_rust_bridge/pull/1325/commits/7c4c48b1497998258078dba5d2ab92062d3286c3)
this compile issue arrises, which is weird since there is no error messages
image

**Comment by fzyzcjy** [comment]
Try to run `dart format` manually. btw `output=none` may suppress the outputs?

**Comment by SiongSng** [comment]
I have successfully made most of the functions effective, but some behaviors on the web seem to be different from io, causing the integration test execution to fail.

Web integration test log:
https://github.com/fzyzcjy/flutter_rust_bridge/actions/runs/5902491055/job/16010626720?pr=1325#step:12:1183

**Comment by JustSimplyKyle** [comment]
Now there's an issue with thiserror.
If we use a error type without the `#[from]` attribute, it works, but if we do?
image
it fails. It's an interesting problem since without the `#[from]` attribute, thiserror is kinda hard to use

**Comment by SiongSng** [comment]
Now all the issues except Valgrind have been fixed, the reason is seems to be related to `panic`

**Comment by JustSimplyKyle** [comment]
Something weird... happen
this is [fine](https://github.com/fzyzcjy/flutter_rust_bridge/actions/runs/5868452973/job/15911323953)
while this [breaks](https://github.com/fzyzcjy/flutter_rust_bridge/actions/runs/5909307658/job/16029737370)!!
which is soo weird since the breaking commit is a contributor one

**Comment by fzyzcjy** [comment]
Great job! So maybe we should fix Valgrind before having a review?

**Comment by SiongSng** [comment]
The issues with Valgrind doesn't seem to be related to this PR, since it's broken since this commit https://github.com/fzyzcjy/flutter_rust_bridge/commit/2e01d4aec8a28dba89de02e0e2d06a2ccf771df3

**Comment by fzyzcjy** [comment]
Ok, then let me rerun it...

File: frb_codegen/src/parser/ty.rs [hunk 1]
```diff
@@ -331,18 +333,21 @@ impl<'a> TypeParser<'a> {
...
                                 .unwrap_or(false),
+                            is_exception: false,
                         }))
                     }
 
-                    [(name, None)] if self.src_enums.contains_key(&name.to_string()) => {
+                    [(name, _)] if self.src_enums.contains_key(&name.to_string()) => {
                         let ident_string = name.to_string();
                         if self.parsed_enums.insert(ident_string.to_owned()) {
+                            // NOTE: not working
```
Comment (fzyzcjy) [inline frb_codegen/src/parser/ty.rs:343]: 👀 what is happening

File: frb_codegen/src/generator/dart/ty_enum.rs [hunk 2]
```diff
@@ -161,6 +162,12 @@ impl TypeDartGeneratorTrait for TypeEnumRefGenerator<'_> {
                 .variants()
                 .iter()
                 .map(|variant| {
+                    let has_backtrace = matches!(&variant.kind, IrVariantKind::Struct(IrStruct {
+                        is_fields_named: true,
+                        fields,
+                        ..
+                    }) if fields.iter().any(|field| field.name.raw == "backtrace"));
```
Comment (fzyzcjy) [inline frb_codegen/src/generator/dart/ty_enum.rs:169]: nit: maybe we should at least extract the string into a constant

File: frb_dart/lib/src/basic.dart [hunk 3]
```diff
@@ -19,6 +19,16 @@ class _DropIdPortGenerator {
...
 
+class PanicException extends FrbException {
+  final String error;
+
+  PanicException(this.error);
+}
+
+PanicException wire2apiPanicError(dynamic raw) {
+  return PanicException(raw as String);
+}
```
Comment (fzyzcjy) [inline frb_dart/lib/src/basic.dart:30]: nit: move PanicException to the sibling of FrbException etc

File: frb_dart/lib/src/basic.dart [hunk 4]
```diff
@@ -109,24 +121,37 @@ abstract class FlutterRustBridgeBase<T extends FlutterRustBridgeWireBase> {
...
+        if (parseErrorData != null) {
+          throw parseErrorData(raw[1]);
+        }
+        throw Exception("tried to parse error data but function is null");
+      case _RUST2DART_ACTION_PANIC:
+        assert(raw.length == 2);
+        if (parsePanicData != null) {
+          throw parsePanicData(raw[1]);
+        }
+        throw Exception("tried to parse panic data but function is null");
```
Comment (fzyzcjy) [inline frb_dart/lib/src/basic.dart:154]: nit: there seems to be some duplicate among them. Maybe we can avoid it?

File: frb_dart/lib/src/platform_independent.dart [hunk 5]
```diff
@@ -113,3 +114,9 @@ bool _listEquals<T>(List<T>? a, List<T>? b) {
   }
   return true;
 }
+
+class FrbAnyhowException implements FrbException {
+  String anyhow;
```
Comment (fzyzcjy) [inline frb_dart/lib/src/platform_independent.dart:119]: nit

```suggestion
  final String anyhow;
```

**Review by fzyzcjy — APPROVED** [review]
Good job! Ready to merge after some nits

File: frb_codegen/src/parser/ty.rs (hunk 1 above)
Comment (JustSimplyKyle) [reply frb_codegen/src/parser/ty.rs:343]: https://github.com/fzyzcjy/flutter_rust_bridge/pull/1325/commits/83a78894084d0e81fd0f40647b125d745a5d0adc#diff-cf4e6088e3742e99ecc674bc551f63661342a4550589d29f5a387eddc772e09d
this part should be fixed in this commit

**Comment by SiongSng** [comment]
All the nits have been fixed.

**Comment by fzyzcjy** [comment]
Looks great! I will check soon

**Comment by fzyzcjy** [comment]
Related: https://github.com/fzyzcjy/flutter_rust_bridge/issues/1327

**Comment by fzyzcjy** [comment]
As mentioned there, the Valgrind issue blocks this PR from merging, but I am really busy and have to squeeze some time to review this PR, let alone time for fixing that. Do you have some interest in digging a bit further about that error :)

**Comment by SiongSng** [comment]
I'm also looking into this issue, I guess it's an environment change that caused the Valgrind test to fail, not directly related to these two PRs.

**Comment by fzyzcjy** [comment]
Totally agree, since the failing commit is completely unrelated

**Comment by dbsxdbsx** [comment]
> I'm also looking into this issue, I guess it's an environment change that caused the Valgrind test to fail, not directly related to these two PRs.

I guess so, because [mine ](https://github.com/fzyzcjy/flutter_rust_bridge/pull/1145#issuecomment-1685888486) also failed recently.

**Comment by fzyzcjy** [comment]
I would like to take the risk and firstly merge this PR, and then fix Valgrind later... Since our Valgrind does not check a lot currently (since even simplest hello world dart program triggers invalid memory access so that part is disabled)

**Comment by fzyzcjy** [comment]
@all-contributors please add @SiongSng for code

**Comment by fzyzcjy** [comment]
@all-contributors please add @JustSimplyKyle for code

**Pull request merged**